File: test/user_agent_identity_shim_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:webspace/services/user_agent_classifier.dart';
import 'package:webspace/services/user_agent_identity_shim.dart';

const _chromeAndroid =
    'Mozilla/5.0 (Linux; Android 14; Pixel 8) AppleWebKit/537.36 '
    '(KHTML, like Gecko) Chrome/120.0.0.0 Mobile Safari/537.36';
const _safariMac =
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 '
    '(KHTML, like Gecko) Version/17.4 Safari/605.1.15';
const _criOs =
    'Mozilla/5.0 (iPhone; CPU iPhone OS 17_4 like Mac OS X) '
    'AppleWebKit/605.1.15 (KHTML, like Gecko) CriOS/120.0.0.0 '
    'Mobile/15E148 Safari/604.1';

void main() {
  group('inferUaEngine', () {
    test('Gecko: Firefox desktop + Android', () {
      expect(inferUaEngine(firefoxLinuxDesktopUserAgent), UaEngine.gecko);
      expect(inferUaEngine(buildFirefoxAndroidUserAgent('152.0')),
          UaEngine.gecko);
    });

    test('WebKit: Safari, FxiOS, and CriOS (iOS is always WebKit)', () {
      expect(inferUaEngine(_safariMac), UaEngine.webkit);
      expect(inferUaEngine(buildFirefoxIosUserAgent('152.0')), UaEngine.webkit);
      // CriOS carries no "Chrome/" token and is iOS => WebKit, not Blink.
      expect(inferUaEngine(_criOs), UaEngine.webkit);
    });

    test('Blink: Chrome for Android and desktop', () {
      expect(inferUaEngine(_chromeAndroid), UaEngine.blink);
      expect(
          inferUaEngine(
              'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 '
              '(KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'),
          UaEngine.blink);
    });

    test('unknown for empty / unrecognized', () {
      expect(inferUaEngine(null), UaEngine.unknown);
      expect(inferUaEngine(''), UaEngine.unknown);
      expect(inferUaEngine('curl/8.0'), UaEngine.unknown);
    });
  });

  group('buildUserAgentIdentityShim', () {
    test('returns null when the engine is unclassifiable', () {
      expect(buildUserAgentIdentityShim(''), isNull);
      expect(buildUserAgentIdentityShim('curl/8.0'), isNull);
    });

    test('Gecko mobile (Firefox-Android): empty vendor, gecko productSub, '
        'frozen oscpu/buildID, mobile platform', () {
      final s = buildUserAgentIdentityShim(buildFirefoxAndroidUserAgent('152.0'))!;
      expect(s, contains("def('vendor', \"\");"));
      expect(s, contains("def('vendorSub', '');"));
      expect(s, contains("def('productSub', \"20100101\");"));
      expect(s, contains("def('oscpu', \"Linux armv8l\");"));
      expect(s, contains("def('buildID', '20181001000000');"));
      expect(s, contains("def('platform', \"Linux armv8l\");"));
      expect(s, contains('__ws_ua_identity_shim__'));
      // The builder must NOT append the evaluator-return; the call site does.
      expect(s.trimRight().endsWith('})();'), isTrue);
    });

    test('Gecko desktop: desktop oscpu, and NO platform override '
        '(desktop_mode_shim owns platform)', () {
      final s = buildUserAgentIdentityShim(firefoxLinuxDesktopUserAgent)!;
      expect(s, contains("def('oscpu', \"Linux x86_64\");"));
      expect(s, contains("def('buildID', '20181001000000');"));
      expect(s, isNot(contains("def('platform'")));
    });

    test('Gecko desktop windows/macos oscpu tokens', () {
      expect(buildUserAgentIdentityShim(firefoxWindowsDesktopUserAgent)!,
          contains("def('oscpu', \"Windows NT 10.0; Win64; x64\");"));
      expect(buildUserAgentIdentityShim(firefoxMacosDesktopUserAgent)!,
          contains("def('oscpu', \"Intel Mac OS X 10.15\");"));
    });

    test('WebKit mobile (FxiOS): Apple vendor, webkit productSub, '
        'oscpu/buildID/userAgentData removed, iPhone platform', () {
      final s = buildUserAgentIdentityShim(buildFirefoxIosUserAgent('152.0'))!;
      expect(s, contains("def('vendor', \"Apple Computer, Inc.\");"));
      expect(s, contains("def('productSub', \"20030107\");"));
      expect(s, contains("removeProp('oscpu');"));
      expect(s, contains("removeProp('buildID');"));
      expect(s, contains("removeProp('userAgentData');"));
      expect(s, contains("def('platform', \"iPhone\");"));
    });

    test('Blink mobile (Chrome-Android): Google vendor, webkit productSub, '
        'no oscpu, keeps userAgentData', () {
      final s = buildUserAgentIdentityShim(_chromeAndroid)!;
      expect(s, contains("def('vendor', \"Google Inc.\");"));
      expect(s, contains("def('productSub', \"20030107\");"));
      expect(s, contains("removeProp('oscpu');"));
      expect(s, contains("def('platform', \"Linux armv8l\");"));
      // Blink keeps userAgentData — it must NOT be removed.
      expect(s, isNot(contains("removeProp('userAgentData')")));
    });
  });
}

File: test/js/user_agent_identity_shim.test.js
// Tier 1 — jsdom assertions for the engine-consistent navigator-identity shim
// (lib/services/user_agent_identity_shim.dart, dumped to
// test/js_fixtures/ua_identity/*.js).
//
// The shim forces navigator.vendor / vendorSub / productSub / oscpu /
// buildID / platform to the values the UA's *claimed* engine really emits,
// so a spoofed UA on a mismatched host engine (Gecko UA on iOS WebKit, etc.)
// doesn't leak the real engine. `oscpu` / `buildID` / `userAgentData` are
// presence-sensitive: on the engines that lack them the property must be
// genuinely absent (`in` === false), not defined as undefined.

const test = require('node:test');
const assert = require('node:assert/strict');
const { loadShim } = require('./helpers/load_shim');

const FX_ANDROID = 'ua_identity/firefox_android.js';
const FX_LINUX_DESKTOP = 'ua_identity/firefox_linux_desktop.js';
const FXIOS = 'ua_identity/fxios.js';
const CHROME_ANDROID = 'ua_identity/chrome_android.js';

// --- Gecko mobile (Firefox for Android) ---

test('Firefox-Android: Gecko vendor / productSub', () => {
  const nav = loadShim(FX_ANDROID).window.navigator;
  assert.equal(nav.vendor, '');
  assert.equal(nav.vendorSub, '');
  assert.equal(nav.productSub, '20100101');
});

test('Firefox-Android: Gecko-only oscpu / buildID present with frozen values', () => {
  const dom = loadShim(FX_ANDROID);
  const nav = dom.window.navigator;
  assert.equal('oscpu' in nav, true);
  assert.equal(nav.oscpu, 'Linux armv8l');
  assert.equal('buildID' in nav, true);
  assert.equal(nav.buildID, '20181001000000');
});

test('Firefox-Android: platform is the frozen "Linux armv8l", not the host', () => {
  assert.equal(loadShim(FX_ANDROID).window.navigator.platform, 'Linux armv8l');
});

// --- Gecko desktop (platform owned by desktop_mode_shim, so NOT set here) ---

test('Firefox-desktop: Gecko identity with desktop oscpu, no platform override', () => {
  const dom = loadShim(FX_LINUX_DESKTOP);
  const nav = dom.window.navigator;
  assert.equal(nav.vendor, '');
  assert.equal(nav.productSub, '20100101');
  assert.equal(nav.oscpu, 'Linux x86_64');
  assert.equal(nav.buildID, '20181001000000');
  // No `def('platform', ...)` on the desktop path — desktop_mode_shim owns it.
  const src = require('./helpers/load_shim').readFixture(FX_LINUX_DESKTOP);
  assert.equal(/def\('platform'/.test(src), false);
});

// --- WebKit mobile (Firefox for iOS — Safari-shaped) ---

test('FxiOS: WebKit vendor / productSub', () => {
  const nav = loadShim(FXIOS).window.navigator;
  assert.equal(nav.vendor, 'Apple Computer, Inc.');
  assert.equal(nav.productSub, '20030107');
});

test('FxiOS: oscpu / buildID are ABSENT (in === false), not undefined', () => {
  const nav = loadShim(FXIOS).window.navigator;
  assert.equal('oscpu' in nav, false);
  assert.equal('buildID' in nav, false);
});

test('FxiOS: platform is "iPhone"', () => {
  assert.equal(loadShim(FXIOS).window.navigator.platform, 'iPhone');
});

// --- Blink mobile (Chrome for Android) ---

test('Chrome-Android: Blink vendor / productSub, no oscpu', () => {
  const nav = loadShim(CHROME_ANDROID).window.navigator;
  assert.equal(nav.vendor, 'Google Inc.');
  assert.equal(nav.productSub, '20030107');
  assert.equal('oscpu' in nav, false);
  assert.equal(nav.platform, 'Linux armv8l');
});

test('Chrome-Android: userAgentData is NOT removed (Blink keeps it)', () => {
  // The shim removes userAgentData only for Gecko/WebKit; the Blink fixture
  // must not carry a removeProp('userAgentData') line.
  const src = require('./helpers/load_shim').readFixture(CHROME_ANDROID);
  assert.equal(/removeProp\('userAgentData'\)/.test(src), false);
});

test('FxiOS DOES remove userAgentData (WebKit lacks it)', () => {
  const src = require('./helpers/load_shim').readFixture(FXIOS);
  assert.equal(/removeProp\('userAgentData'\)/.test(src), true);
});

// --- Detection hardening ---

test('identity getters land on Navigator.prototype, not the instance', () => {
  const dom = loadShim(FX_ANDROID);
  const own = Object.getOwnPropertyNames(dom.window.navigator);
  for (const leaked of ['vendor', 'vendorSub', 'productSub', 'oscpu',
                        'buildID', 'platform']) {
    assert.equal(own.includes(leaked), false,
      `${leaked} leaks as own-property: ${JSON.stringify(own)}`);
  }
});

test('identity getters stringify as [native code]', () => {
  const dom = loadShim(FX_ANDROID);
  const desc = Object.getOwnPropertyDescriptor(
    dom.window.Navigator.prototype, 'vendor');
  const s = dom.window.Function.prototype.toString.call(desc.get);
  assert.match(s, /\[native code\]/);
});

test('shim loads cleanly and is idempotent under jsdom', () => {
  const dom = loadShim(FX_ANDROID);
  assert.doesNotThrow(() =>
    dom.window.eval(require('./helpers/load_shim').readFixture(FX_ANDROID)));
  assert.equal(dom.window.navigator.vendor, '');
});
